File: frameworks/payload.core/src/main/java/com/qomolangma/frameworks/payload/core/OneElement.java
package com.qomolangma.frameworks.payload.core;

import java.util.Map;

public class OneElement<D> {
    private final ConvertRules<D> conventions;

    @SafeVarargs
    public OneElement(ConvertRule<D>... conventions) {
        this.conventions = new ConvertRules<>(conventions);
    }

    @SafeVarargs
    public static <D> OneElement<D> one(ConvertRule<D>... conventions) {
        return new OneElement<>(conventions);
    }

    public Map<String, Object> transform(D d) {
        return conventions.convert(d);
    }
}

File: frameworks/payload.core/src/main/kotlin/com/qomolangma/frameworks/payload/core/OneElement.kt
package com.qomolangma.frameworks.payload.core

class OneElement<D> constructor(vararg conventions: ConvertRule<D>) {
    private val conventions: ConvertRules<D>

    init {
        this.conventions = ConvertRules(*conventions)
    }

    fun transform(d: D): Map<String, Any> {
        return conventions.convert(d)
    }

    companion object {
        @JvmStatic
        fun <D> one(vararg conventions: ConvertRule<D>): OneElement<D> {
            return OneElement(*conventions)
        }
    }
}
